internal/docker/container.go

```diff
@@ -334,3 +356,61 @@ func (r *ContainerRunner) runSuite(options containerStartOptions) (containerID s
 		options.Environment)
 	return
 }
+
+// registerInterruptOnSignal runs tearDown on SIGINT / Interrupt.
+func (r *ContainerRunner) registerInterruptOnSignal(containerID, suiteName string, cleanedUp *bool, interrupted *bool) chan os.Signal {
+	sigChan := make(chan os.Signal)
+	signal.Notify(sigChan, os.Interrupt)
+
+	go func(c <-chan os.Signal, cleanedUp, interrupted *bool, containerID, suiteName string) {
+		sig := <-c
+		if sig == nil {
+			return
+		}
+		log.Info().Str("suite", suiteName).Msg("Interrupting suite")
+		*interrupted = true
+		r.tearDown(containerID, suiteName, cleanedUp)
+	}(sigChan, cleanedUp, interrupted, containerID, suiteName)
+	return sigChan
+}
+
+// registerSkipSuitesOnSignal prevent new suites from being executed when a SIGINT is captured.
+func (r *ContainerRunner) registerSkipSuitesOnSignal() chan os.Signal {
+	sigChan := make(chan os.Signal)
+	signal.Notify(sigChan, os.Interrupt)
+
+	go func(c <-chan os.Signal, cr *ContainerRunner) {
+		for {
+			sig := <-c
+			if sig == nil {
+				return
+			}
+			if cr.interrupted {
+				os.Exit(1)
+			}
+			log.Info().Msg("Ctrl-C captured. Ctrl-C again to exit now.")
+			cr.interrupted = true
+		}
+	}(sigChan, r)
+	return sigChan
+}
+
+// unregisterSignalCapture remove the signal hook associated to the chan c.
+func unregisterSignalCapture(c chan os.Signal) {
+	signal.Stop(c)
+	close(c)
+}
+
+// tearDown stops the test environment and remove docker containers.
+func (r *ContainerRunner) tearDown(containerID, suiteName string, done *bool) {
```

Comment:
The pointer allow to share the value between line L372 and L337.
This is to prevent the tearDown to be runned twice (and generate errors due to already deleted docker container).

As this is not possible to "undo" a defer, the is the only workaround I get.

Reply:
Deleting an already removed container **shouldn't** cause any errors, but if it does, we should address that.
This would allow for a simpler cleanup logic over here :)

Reply:
Actually it is not, the error pops out only when the 2 calls are so close. So when the removal from the ^C capture is still in progress and `defer` kicks in, there is an error:
```
23:50:13 ERR Failed to tear down environment error="Error response from daemon: removal of container bde15b4fed228f21407dc4a83c3d5c2ed34e2fb7a964c915b9a427aac513c491 is already in progress" suite="saucy chrome test-1"
```

Reply:
Aye, better to ingore the error if we get that specific one (since we know the container is being dealt with) 👍

Reply:
We are good to go :-)